Lib9c/Action/Guild/Migration/Controls/GuildMigrationCtrl.cs

```diff
@@ -0,0 +1,75 @@
+using Bencodex.Types;
+using Libplanet.Action.State;
+using Nekoyume.Extensions;
+using Nekoyume.Model.State;
+using Nekoyume.Module;
+using Nekoyume.Module.Guild;
+using Nekoyume.PolicyAction.Tx.Begin;
+using Nekoyume.TypedAddress;
+using Serilog;
+
+namespace Nekoyume.Action.Guild.Migration.Controls
+{
+    public static class GuildMigrationCtrl
+    {
+        /// <summary>
+        /// Migrate the pledge to the guild if the <paramref name="target"/> has contracted pledge
+        /// with Planetarium (<see cref="MeadConfig.PatronAddress"/>).
+        /// </summary>
+        /// <param name="world"></param>
+        /// <param name="target"></param>
+        /// <returns></returns>
+        public static (IWorld World, bool ShouldFail) MigratePlanetariumPledgeToGuild(IWorld world, AgentAddress target)
+        {
+            var logger = Log.ForContext(typeof(GuildMigrationCtrl));
+
+            var planetariumPatronAddress = new AgentAddress(MeadConfig.PatronAddress);
+            if (world.GetJoinedGuild(planetariumPatronAddress) is not { } planetariumGuildAddress)
+            {
+                logger.Warning("Planetarium seems not to make guild yet. Skip auto joining.");
+                return (world, true);
+            }
+
+            if (!world.TryGetGuild(planetariumGuildAddress, out var planetariumGuild))
+            {
+                logger.Error(
+                    "Planetarium address seems to join guild but it failed to fetch " +
+                    "the guild. It seems a bug situation. It skips auto joining but " +
+                    "you must investigate this issue.");
+                return (world, true);
+            }
+
+            if (planetariumGuild.GuildMasterAddress != planetariumPatronAddress)
+            {
+                logger.Error(
+                    "Planetarium address seems to join guild but it is the owner of the guild." +
+                    "It seems a bug situation. It skips auto joining but " +
+                    "you must investigate this issue.");
+                return (world, true);
+            }
+
+            if (world.GetJoinedGuild(target) is { } joinedGuildAddress)
+            {
+                Log.ForContext<AutoJoinGuild>()
+                    .Verbose(
+                        "{Signer} is already joined to {Guild}. Skip auto joining.",
+                        target,
+                        joinedGuildAddress);
+                return (world, true);
+            }
+
+            var pledgeAddress = target.GetPledgeAddress();
+
+            // Patron contract structure:
+            // [0] = PatronAddress
+            // [1] = IsApproved
+            // [2] = Mead amount to refill.
+            if (!world.TryGetLegacyState(pledgeAddress, out List list) || list.Count < 3 || list[0].ToAddress() != MeadConfig.PatronAddress)
+            {
+                return (world, true);
```

Comment:
In general (TryParse, TryGetValue, etc.), if the return type is bool, there should be a guarantee that this method will not throw any exceptions.
I think it would be a bit confusing for someone using this method to have a mix of successes, failures, and exceptions.
What I would suggest is to define and throw an exception class that can be used in the case of returning failure (true).
Then, in AutoJoinGuild, make sure that only that exception is handled by the catch.

Reply:
So are you saying that we should create an exception for every reason why it fails and throw it? If so, I think I'd need about 6 exception classes. I'm a little torn because I don't think there's anything extra to handle when I get an exception, or maybe it's not very C#-like to return something like (result, err) like Go, so I could fix it with something like `TryMigrate...`?

Reply:
I tried to apply your opinion at [b4d361e](https://github.com/planetarium/lib9c/pull/2722/commits/b4d361e6d590aba3e87ae990fbd21b2ab00649f0). Please review again 🙏🏻

Reply:
LGTM 😀